flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java

```diff
@@ -1542,6 +1542,21 @@ void testUsingAddJar() throws Exception {
                 "drop function lowerUdf");
     }
 
+    @Test
+    void testUsingAddJarWithCheckpointing() throws Exception {
+        env().enableCheckpointing(100);
+        tEnv().executeSql(String.format("ADD JAR '%s'", jarPath));
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+        testUserDefinedFunctionByUsingJar(
+                env ->
+                        env.executeSql(
+                                String.format(
+                                        "create function lowerUdf as '%s' LANGUAGE JAVA",
+                                        udfClassName)),
+                "drop function lowerUdf");
+        assertThat(contextClassLoader.equals(Thread.currentThread().getContextClassLoader()));
+    }
```

Comment:
What does exactly this test checking?
I'm asking since I executed it separately without changes in `TableEnvironmentImpl`, `StreamTableEnvironment`

Reply:
@snuyanzin Thanks for reviewing. This PR is targeting the 1.20 release branch. If you try executing this test case in 1.20 branch, you can see this failing. However this issue is currently not occurring in the master branch due to a probable refactoring around the job submission area for the 2.0 versions. Please see the comments captured in a related ticket here: https://issues.apache.org/jira/browse/FLINK-36065

Reply:
yep, I was checking in 2.0/master

thanks for clarifying this.

Then The question would be why are we creating something new here instead of backporting existing solution?

Reply:
also regarding test
this line doesn't make any sense
```java
assertThat(contextClassLoader.equals(Thread.currentThread().getContextClassLoader()));
```
since you put into assertThat `contextClassLoader.equals(Thread.currentThread().getContextClassLoader())` and no assert method was invoked

Reply:
> Btw are you sure that the issue was solved within [FLINK-36065](https://issues.apache.org/jira/browse/FLINK-36065)? Based on comments there it is not that clear

Checking the git logs on the related files for this issue, I could find only this PR which is updating the logic around compilation of jobgraph from the stream graph within the Job manager. 

> Then The question would be why are we creating something new here instead of backporting existing solution?

I assume this issue fix in 2.0 is a side effect of the refactoring and not a direct fix. As confirmed by @JunRuiLee on the https://issues.apache.org/jira/browse/FLINK-36065 they have no plans to backport this refactoring to lower versions. So I sticked onto this approach.

Reply:
Also, updated the test case with the right assertion.

Reply:
I'm still puzzled why we are going here in different direction in compare with master branch where same code for TableEnvironmentImpl was not change and the problem is solved. Having this in mind I'm not sure this approach the correct one... 
>I assume this issue fix in 2.0 is a side effect of the refactoring and not a direct fix. 

Did you do git bisect or on what is this assumption based? I'm asking since I haven't found any confirmation that this issue was fixed by exactly commits within [FLINK-36065](https://issues.apache.org/jira/browse/FLINK-36065)


Also it would be great to get some responses for the comment at https://github.com/apache/flink/pull/25656#discussion_r1894239482

Reply:
I haven't performed a git bisect while analysis and it was assumed by checking the implementation done for [FLINK-36065](https://issues.apache.org/jira/browse/FLINK-36065). 

But now I performed a test on 2 different commit heads using this test case for confirmation: 
1.  https://github.com/apache/flink/commit/98f712640555333d5e1842a53441b283ca2f6e29 (before FLINK-36065) - test case fails.
2. https://github.com/apache/flink/commit/64cbcd2d4b360f03f7964edece6f3211d184fb06  (after FLINK-36065) - test case succeeds.

<img width="1275" alt="image" src="https://github.com/user-attachments/assets/fee2f4d5-8a47-48bd-8546-2363fbd51cc2" />

Reply:
@snuyanzin I believe the key point here is that the changes that fix the issue for 2.0 have a much broader scope and it has been agreed that we will not backport features from 2.x into 1.x line (as discussed [here](https://github.com/apache/flink/pull/25656#discussion_r1852573652)), hence a more narrow fix is required.